### PR TITLE
Ncs 83 status eligibility check

### PR DESCRIPTION
**Pull request by ICoxCH**

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidation.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+public class CompanyStatusValidation implements EligibilityRule<CompanyProfileApi> {
+
+    private final Set<String> allowedStatuses;
+
+    public CompanyStatusValidation(Set<String> allowedStatuses) {
+        this.allowedStatuses = allowedStatuses;
+    }
+
+    @Override
+    public void validate(CompanyProfileApi input) throws EligibilityException {
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidation.java:19]: are you able to rename input to api? just think it would read better

Comment (markpit) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidation.java:19]: shouldn't we be validating the company profile and not the api?

File: src/test/java/uk/gov/companieshouse/confirmationstatementapi/contoller/ConfirmationStatementControllerTest.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import uk.gov.companieshouse.api.model.transaction.Transaction;
+import uk.gov.companieshouse.confirmationstatementapi.exception.ServiceException;
+import uk.gov.companieshouse.confirmationstatementapi.service.ConfirmationStatementService;
+
+import java.io.IOException;
+import java.net.URI;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/confirmationstatementapi/contoller/ConfirmationStatementControllerTest.java:17]: * on import

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidation.java (hunk 1 above)
Comment (markpit) [reply src/main/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidation.java:19]: ok so CompanyProfileApi is returned from the sdk but is actually the company profile

File: src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    public ResponseEntity<Object> createConfirmationStatement(Transaction transaction) throws ServiceException {
+        var companyProfile = companyProfileService.getCompanyProfile(transaction.getCompanyNumber());
+
+        try {
+            for (EligibilityRule<CompanyProfileApi> eligibilityRule : eligibilityRules) {
+                eligibilityRule.validate(companyProfile);
+            }
+        } catch (EligibilityException e) {
+            return ResponseEntity.badRequest().body(e.getEligibilityFailureReason());
```
Comment (markpit) [inline src/main/java/uk/gov/companieshouse/confirmationstatementapi/service/ConfirmationStatementService.java:38]: i think it might be good to log.info if a company is ineligible and log the code

File: src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidationTest.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    @Test
+    void validateDoesNotThrowOnAllowedValue() throws EligibilityException {
+        CompanyProfileApi companyProfileApi = new CompanyProfileApi();
+        companyProfileApi.setCompanyStatus(ALLOWED_VALUE);
+
+        companyStatusValidation.validate(companyProfileApi);
+    }
+
+    @Test
+    void validateThrowsOnDisallowedValue() throws EligibilityException {
```
Comment (markpit) [inline src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidationTest.java:36]: i'd also add a test to pass null into the rule,  nulls/undefineds can cause issues

Comment (markpit) [reply src/test/java/uk/gov/companieshouse/confirmationstatementapi/eligibility/impl/CompanyStatusValidationTest.java:36]: or rather, the company status set to null, maybe do both!

**Review by markpit — APPROVED** [review]
